### PR TITLE
More performant newline splitting

### PR DESCRIPTION
**Pull request by nietaki**
The issue popped up while using `server_sent_event` that the service using it spent a lot of the time running regular expression code.

It seems like it was caused by splitting binaries by newlines using regular expressions. The root cause is that even if you specify `parts: 2` in the `split` function options, if you're splitting by a regular expression, the whole string is going to be processed instead of just the part you need.

See relevant code:

- https://github.com/elixir-lang/elixir/blob/v1.6.4/lib/elixir/lib/string.ex#L383-L385
- https://github.com/elixir-lang/elixir/blob/v1.6.4/lib/elixir/lib/regex.ex#L472-L477

This approach should be much faster, as per the benchmark results:

```
Name                                 ips        average  deviation         median         99th %
with_binary_split_compiled    10114.35 K      0.0989 μs   ±177.83%      0.0900 μs       0.160 μs
with_binary_split               806.20 K        1.24 μs   ±114.65%        1.20 μs        2.10 μs
with_lazy_compiled_pattern      717.57 K        1.39 μs    ±88.91%        1.30 μs        2.40 μs
with_elixir_string_split        669.83 K        1.49 μs    ±76.74%        1.40 μs        2.60 μs
with_erlang_regex                 0.43 K     2350.80 μs     ±3.82%        2342 μs     2697.22 μs
with_regex                        0.37 K     2698.95 μs     ±5.03%        2651 μs     3193.58 μs

Comparison: 
with_binary_split_compiled    10114.35 K
with_binary_split               806.20 K - 12.55x slower
with_lazy_compiled_pattern      717.57 K - 14.10x slower
with_elixir_string_split        669.83 K - 15.10x slower
with_erlang_regex                 0.43 K - 23776.81x slower
with_regex                        0.37 K - 27298.11x slowe
```

## Notes

- More performance gains are possible by moving from `String.split` to `:binary.split`, it could introduce some confusion to an otherwise consistent Elixir'y code
- Further performance gains are possible using compiling the patterns, (see the benchmark), but because it's not possible to store the compiled pattern as a module attribute, it would lead to much uglier code, at least using the approaches I thought about so far

### CODE REVIEW

File: test/server_sent_event/splitting_test.exs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  @new_line_regex ~r/\R/
+  @new_line_sequences ["\r\n", "\r", "\n"]
+
+  def with_regex(stream) do
+    String.split(stream, @new_line_regex, parts: 2)
+  end
+
+  def with_binary_split(stream) do
+    # not global means just two parts
```
Comment (CrowdHailer) [inline test/server_sent_event/splitting_test.exs]: What does global mean here?

Comment (nietaki) [reply test/server_sent_event/splitting_test.exs]: http://erlang.org/doc/man/binary.html#split-3 - one of the options for the function. I'll put it in the docs.

**Comment by nietaki** [comment]
I added two more approaches to the benchmark, to make sure we covered all the cases:

- trying to manipulate the erlang `:re` interface to do more efficient splitting into 2 parts
- using a named ets table to store the compiled pattern

The takeaway is I can't make `:re` do a more efficient approach to splitting and the overhead of a named ets table makes it not worth it.

Basically unless you compile the pattern locally and pass it around, there's no real performance gains.

**Comment by CrowdHailer** [comment]
```
Comparison: 
with_binary_split_compiled     7677.31 K
with_lazy_compiled_pattern      813.37 K - 9.44x slower
with_binary_split               742.23 K - 10.34x slower
with_elixir_string_split        518.96 K - 14.79x slower
with_erlang_regex                 0.30 K - 25979.53x slower
with_regex                        0.27 K - 28526.78x slower
```
So top one not an option. I think the ets addition does complicate things, I would suggest binary split?

Also, at least erlang regex is of same order of magnitude as Elixir regex.

**Comment by CrowdHailer** [comment]
Also the failed test is a classic, formatter.

**Review by CrowdHailer — APPROVED** [review]

**Comment by josevalim** [comment]
Note `String.split/2` does accept a compiled pattern too but since you want to break it exactly in two parts, then I would go with `:binary.split/2` indeed (it should not make a big difference but alas). I'm mentioning it just for completeness. :+1:

**Pull request merged**